edgexfoundry/app-service-configurable
# refactor(snap): edgex-snap-hooks related upgrade

This PR upgrades snap hooks to [2.4.1 ](https://github.com/canonical/edgex-snap-hooks/releases/tag/v2.4.1)with new function and refactor:
- Combine install hook and configure hook
- Use library function for copying directory
- Remove export bash command to process valid .env files
- Use log package for logging
- Use library function for autostart processing

Signed-off-by: Mengyi Wang <mengyi.wang@canonical.com>







**If your build fails**  due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/app-service-configurable/blob/main/.github/CONTRIBUTING.md

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] I have fully tested (add details below) this the new feature or bug fix (if not, why?)  [app-service-configurable test suite](https://github.com/canonical/edgex-snap-testing/tree/main/test/suites/app-service-configurable) passed with `FULL_CONFIG_TEST=true`
  <link to docs PR>

## Testing Instructions

1. Install and validate that all config files have been copied in right place:
```
tree /var/snap/edgex-app-service-configurable/current/config
.
/var/snap/edgex-app-service-configurable/current/config
└── res
    ├── external-mqtt-trigger
    │   └── configuration.toml
    ├── functional-tests
    │   └── configuration.toml
    ├── http-export
    │   └── configuration.toml
    ├── metrics-influxdb
    │   └── configuration.toml
    ├── mqtt-export
    │   └── configuration.toml
    ├── push-to-core
    │   └── configuration.toml
    ├── rules-engine
    │   └── configuration.toml
    └── sample
        └── configuration.toml

9 directories, 8 files
```
2. Setup and start:
```
sudo snap set edgex-app-service-configurable app-options=true
sudo snap set edgex-app-service-configurable config.writable-loglevel=DEBUG
sudo snap set edgex-app-service-configurable config.service-startupmsg="#test start up message#"
sudo snap start edgex-app-service-configurable
```
3. Check the .env file is valid now without `export` command:
```
$ cat /var/snap/edgex-app-service-configurable/current/config/res/app-service-configurable.env

WRITABLE_LOGLEVEL="DEBUG"
SERVICE_STARTUPMSG="#test start up message#"
```
4. Validate env vars in `.env` have been exported successfully:
```
$ sudo snap logs -f edgex-app-service-configurable

msg="#test start up message#"
level=DEBUG msg="xxx"
```
5. Validate autostart via snap option works:
```
$ sudo snap set edgex-app-service-configurable autostart=true
$ snap services edgex-app-service-configurable
Service                                                  Startup  Current  Notes
edgex-app-service-configurable.app-service-configurable  enabled  active   -

$ sudo snap set edgex-app-service-configurable autostart=false
$ snap services edgex-app-service-configurable
Service                                                  Startup   Current   Notes
edgex-app-service-configurable.app-service-configurable  disabled  inactive  -
```

## New Dependency Instructions (If applicable)
